File: mtgSearch/src/main/java/com/dbottillo/mtgsearchfree/view/activities/CardsActivity.java
package com.dbottillo.mtgsearchfree.view.activities;

import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.view.PagerTabStrip;
import android.support.v4.view.ViewPager;
import android.widget.RelativeLayout;
import android.widget.Toast;

import com.dbottillo.mtgsearchfree.R;
import com.dbottillo.mtgsearchfree.base.MTGApp;
import com.dbottillo.mtgsearchfree.model.CardsBucket;
import com.dbottillo.mtgsearchfree.presenter.CardsPresenter;
import com.dbottillo.mtgsearchfree.resources.Deck;
import com.dbottillo.mtgsearchfree.resources.MTGCard;
import com.dbottillo.mtgsearchfree.resources.MTGSet;
import com.dbottillo.mtgsearchfree.util.MaterialWrapper;
import com.dbottillo.mtgsearchfree.util.UIUtil;
import com.dbottillo.mtgsearchfree.view.CardsView;

import java.util.ArrayList;
import java.util.Arrays;

import butterknife.Bind;
import butterknife.ButterKnife;

public class CardsActivity extends CommonCardsActivity implements CardsView, ViewPager.OnPageChangeListener {

    private static final String KEY_SEARCH = "Search";
    private static final String KEY_SET = "Set";
    private static final String KEY_DECK = "Deck";
    private static final String POSITION = "Position";

    private MTGSet set = null;
    private Deck deck = null;
    private String search = null;
    private int startPosition = 0;
    private CardsBucket bucket;

    @Bind(R.id.cards_view_pager)
    ViewPager viewPager;
    @Bind(R.id.cards_tab_strip)
    PagerTabStrip pagerTabStrip;
    @Bind(R.id.card_add_to_deck)
    FloatingActionButton fabButton;

    CardsPagerAdapter adapter;

    CardsPresenter cardsPresenter;

    public void onCreate(Bundle bundle) {
        super.onCreate(bundle);
        setContentView(R.layout.activity_cards);

        ButterKnife.bind(this);

        setupView();

        MTGApp.dataGraph.inject(this);
        cardsPresenter.init(this);

        if (getIntent() != null) {
            if (getIntent().hasExtra(KEY_SET)) {
                set = getIntent().getParcelableExtra(KEY_SET);
                getSupportActionBar().setTitle(set.getName());

            } else if (getIntent().hasExtra(KEY_SEARCH)) {
                search = getIntent().getStringExtra(KEY_SEARCH);

            } else if (getIntent().hasExtra(KEY_DECK)) {
                deck = getIntent().getParcelableExtra(KEY_DECK);
            }
        }
        startPosition = getIntent().getIntExtra(POSITION, 0);

        cardsPresenter.loadIdFavourites();

    }

    public void setupView() {
        setupToolbar();
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        MaterialWrapper.setElevation(toolbar, 0f);

        pagerTabStrip.setTabIndicatorColor(getResources().getColor(R.color.white));
        pagerTabStrip.setBackgroundColor(getResources().getColor(R.color.color_primary));
        pagerTabStrip.setTextColor(getResources().getColor(R.color.white));
        RelativeLayout.LayoutParams par = (RelativeLayout.LayoutParams) fabButton.getLayoutParams();
        if (isPortrait) {
            par.addRule(RelativeLayout.CENTER_HORIZONTAL, RelativeLayout.TRUE);
        } else {
            par.addRule(RelativeLayout.ALIGN_PARENT_RIGHT, RelativeLayout.TRUE);
            par.rightMargin = UIUtil.dpToPx(this, 16);
        }
        fabButton.setLayoutParams(par);
        viewPager.addOnPageChangeListener(this);
    }

    public String getPageTrack() {
        if (deck != null) {
            return "/deck";
        }
        return "/cards";
    }

    private void reloadAdapter() {
        boolean showImage = getSharedPreferences().getBoolean(BasicFragment.Companion.getPREF_SHOW_IMAGE(), true);
        adapter = new CardsPagerAdapter(this, deck != null, showImage, bucket.getCards());
        viewPager.setAdapter(adapter);
        viewPager.setCurrentItem(startPosition);
        updateMenu();
    }

    public void favClicked() {
        MTGCard currentCard = adapter.getItem(viewPager.getCurrentItem());
        if (Arrays.asList(idFavourites).contains(currentCard.getMultiVerseId())) {
            cardsPresenter.removeFromFavourite(currentCard);
        } else {
            cardsPresenter.saveAsFavourite(currentCard);
        }
    }

    public MTGCard getCurrentCard() {
        return adapter.getItem(viewPager.getCurrentItem());
    }

    public void toggleImage(boolean show) {
        reloadAdapter();
    }

    public void favIdLoaded(int[] favourites) {
        idFavourites = favourites;

        if (adapter == null) {
            // first time needs to load cards
            if (set != null) {
                cardsPresenter.loadCards(set);
            } else if (search != null) {
                // load search
            } else {
                // something very bad happened here
                throw new UnsupportedOperationException();
            }
        } else {
            updateMenu();
        }
    }

    public void luckyCardsLoaded(ArrayList<MTGCard> cards) {
        throw new UnsupportedOperationException();
    }

    public void cardLoaded(CardsBucket bucket) {
        this.bucket = bucket;
        reloadAdapter();
    }

    public void showError(String message) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
    }

    public void onPageScrollStateChanged(int state) {
        if (state == ViewPager.SCROLL_STATE_IDLE) {
            setFabScale(1.0f);
            updateMenu();
        }
    }

    public void onPageSelected(int position) {
        updateMenu();
    }

    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
        if (positionOffset != 0.0) {
            if (positionOffset < 0.5) {
                setFabScale(1.0f - positionOffset);
            } else {
                setFabScale(positionOffset);
            }
        }
    }

    public void setFabScale(float value) {
        fabButton.setScaleX(value);
        fabButton.setScaleY(value);
    }

}

File: feature_cards/src/main/kotlin/com/dbottillo/mtgsearchfree/cards/CardsActivity.kt
package com.dbottillo.mtgsearchfree.cards

import android.graphics.Bitmap
import android.graphics.Point
import android.net.Uri
import android.os.Bundle
import android.view.View
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.tabs.TabLayout
import androidx.viewpager.widget.ViewPager
import android.widget.RelativeLayout
import android.widget.Toast
import com.dbottillo.mtgsearchfree.model.CardsCollection
import com.dbottillo.mtgsearchfree.model.MTGCard
import com.dbottillo.mtgsearchfree.ui.CommonCardsActivity
import com.dbottillo.mtgsearchfree.ui.views.CardPresenter
import com.dbottillo.mtgsearchfree.ui.views.MTGLoader
import com.dbottillo.mtgsearchfree.util.LOG
import com.dbottillo.mtgsearchfree.util.dpToPx
import com.dbottillo.mtgsearchfree.util.gone
import com.dbottillo.mtgsearchfree.util.setIcon
import com.dbottillo.mtgsearchfree.util.show
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import dagger.android.AndroidInjection
import javax.inject.Inject

class CardsActivity : CommonCardsActivity(), ViewPager.OnPageChangeListener, CardsActivityView {

    private val viewPager by lazy { findViewById<ViewPager>(R.id.cards_view_pager) }
    private val tabLayout by lazy { findViewById<TabLayout>(R.id.cards_tab_layout) }
    private val fabButton by lazy { findViewById<FloatingActionButton>(R.id.card_add_to_deck) }
    private val loader by lazy { findViewById<MTGLoader>(R.id.loader) }
    private val previewBanner by lazy(LazyThreadSafetyMode.NONE) { findViewById<View>(R.id.preview_banner) }

    private var adapter: CardsPagerAdapter? = null

    @Inject lateinit var cardsPresenter: CardsActivityPresenter
    @Inject lateinit var cardPresenter: CardPresenter

    public override val currentCard: MTGCard?
        get() {
            LOG.d()
            return adapter?.getItem(viewPager.currentItem)
        }

    override fun onCreate(bundle: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(bundle)
        setContentView(R.layout.activity_cards)

        fabButton.setOnClickListener {
            LOG.d()
            currentCard?.let { openDialog("add_to_deck", navigator.newAddToDeckFragment(it)) }
        }
        setupView()

        cardsPresenter.init(this, intent)
    }

    private fun setupView() {
        setupToolbar(R.id.toolbar)
        supportActionBar?.let {
            it.setDisplayHomeAsUpEnabled(true)
            it.setDisplayHomeAsUpEnabled(true)
        }
        toolbar.elevation = 0f

        tabLayout.setupWithViewPager(viewPager)
        val par = fabButton.layoutParams as RelativeLayout.LayoutParams
        if (isPortrait) {
            par.addRule(RelativeLayout.CENTER_HORIZONTAL, RelativeLayout.TRUE)
        } else {
            par.addRule(RelativeLayout.ALIGN_PARENT_RIGHT, RelativeLayout.TRUE)
            par.rightMargin = this.dpToPx(16)
        }
        fabButton.layoutParams = par
        viewPager.addOnPageChangeListener(this)

        val display = windowManager.defaultDisplay
        val size = Point()
        display.getSize(size)

        previewBanner.setOnClickListener {
            val dialogBuilder = MaterialAlertDialogBuilder(this)
            dialogBuilder.setMessage(getString(R.string.set_preview_banner_text))
                    .setCancelable(true)
                    .setPositiveButton("Ok") { dialog, _ -> dialog.dismiss() }
            val alert = dialogBuilder.create()
            alert.setTitle(R.string.set_preview_banner)
            alert.show()
        }
    }

    override fun getPageTrack(): String? {
        return if (cardsPresenter.isDeck()) {
            "/deck"
        } else "/cards"
    }

    override fun updateAdapter(cards: CardsCollection, showImage: Boolean, startPosition: Int) {
        LOG.d()
        adapter = CardsPagerAdapter(context = this, showImage = showImage,
                cards = cards, cardPresenter = cardPresenter, trackingManager = trackingManager)
        viewPager.adapter = adapter
        viewPager.currentItem = startPosition
        syncMenu()
    }

    public override fun favClicked() {
        LOG.d()
        cardsPresenter.favClicked(currentCard)
    }

    public override fun toggleImage(show: Boolean) {
        LOG.d()
        cardsPresenter.toggleImage(show)
    }

    override fun syncMenu() {
        cardsPresenter.updateMenu(currentCard)
    }

    override fun showError(errorMessage: String) {
        Toast.makeText(this, errorMessage, Toast.LENGTH_SHORT).show()
    }

    override fun onPageScrollStateChanged(state: Int) {
        if (state == ViewPager.SCROLL_STATE_IDLE) {
            setFabScale(1.0f)
            syncMenu()
        }
    }

    override fun onPageSelected(position: Int) {
        syncMenu()
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
        if (positionOffset.toDouble() != 0.0) {
            if (positionOffset < 0.5) {
                setFabScale(1.0f - positionOffset)
            } else {
                setFabScale(positionOffset)
            }
        }
    }

    private fun setFabScale(value: Float) {
        fabButton.scaleX = value
        fabButton.scaleY = value
    }

    override fun updateTitle(name: String) {
        title = name
    }

    override fun updateTitle(resource: Int) {
        title = getString(resource)
    }

    override fun showFavMenuItem() {
        favMenuItem?.let { it.isVisible = true }
    }

    override fun updateFavMenuItem(text: Int, icon: Int) {
        favMenuItem?.let { fav ->
            fav.title = getString(text)
            fav.setIcon(this, icon, R.attr.colorPrimary)
        }
    }

    override fun hideFavMenuItem() {
        favMenuItem?.let { it.isVisible = false }
    }

    override fun setImageMenuItemChecked(checked: Boolean) {
        imageMenuItem?.isChecked = checked
    }

    override fun showLoading() {
        loader.show()
    }

    override fun hideLoading() {
        loader.gone()
    }

    override fun shareImage(bitmap: Bitmap) {
        currentCard?.let { cardsPresenter.shareImage(bitmap) }
    }

    override fun shareUri(uri: Uri) {
        shareUriArtwork(currentCard?.name ?: "", uri)
    }

    override fun onDestroy() {
        super.onDestroy()
        cardsPresenter.onDestroy()
    }

    override fun renderPreview(preview: Boolean) {
        if (preview) {
            fabButton.hide()
            previewBanner.show()
        } else {
            fabButton.show()
            previewBanner.gone()
        }
    }
}

const val KEY_SEARCH = "Search"
const val KEY_SET = "Set"
const val KEY_FAV = "Fav"
const val KEY_DECK = "Deck"
const val POSITION = "Position"
